README.md
RedSentinel [![Build Status](https://travis-ci.org/raydog/redsentinel.svg?branch=master)](https://travis-ci.org/raydog/redsentinel) [![Coverage Status](https://coveralls.io/repos/github/raydog/redsentinel/badge.svg?branch=master)](https://coveralls.io/github/raydog/redsentinel?branch=master)
===============

A stable, and easy to use redis-sentinel client for Node.js. Features include:

 - Simple, event-based API.
 - Randomized connections to sentinels, and automatic reconnection on sentinel failure.
 - Uses Sentinel's Pub/Sub commands for real-time config updates during failovers.
 - Uses Sentinel info to help dodge outages.

Note: This is a fork of [Vungle/simple-sentinel](https://github.com/Vungle/simple-sentinel), since the original project seems to have been abandoned.

### Installing

```sh
npm install --save redsentinel
```

### Sample Code

```javascript
var RedSentinel = require('redsentinel');

// Put all sentinel servers in here. A random one will be used:
var sentinels = [
  { host: '10.2.3.41', port: 26379 },
  { host: '10.2.3.42', port: 26379 },
  { host: '10.2.3.43', port: 26379 },
  { host: '10.2.3.44', port: 26379 },
  { host: '10.2.3.45', port: 26379 }
];

var options = {
  // Omit this to track all servers that the sentinel knows about:
  watchedNames: ["replica_a", "replica_b", "replica_d"]
};

// Create a sentinel object to track the given Replicas:
var sentinel = new RedSentinel(sentinels, options);

// Keep track of the Masters here:
var masters = {
  replica_a: null,
  replica_b: null,
  replica_d: null
};

// Listen for connection info changes:
sentinel.on('change', function (name, replica) {
  console.log("Just got connection info for Replica:", name, replica.toString());
  masters[name] = replica.connectMaster();
});
```

### API

#### RedSentinel

This object is used to connect to a redis-sentinel cluster. Its constructor is the main thing that's exported. It is an EventEmitter, and will notify you about outages or failovers as they happen.

##### Constructor: (sentinels, config)

Will open the connection to one of the sentinels in the 'sentinels' array.

- `sentinels` is an Array of objects, each with connection info for a single redis-sentinel. These objects should contain:
    - `host` (**String**) is the hostname to connect to.
    - `port` (**Number**) is the port to connect to. Default: 26379.

- `config` is an optional object used to store other configuration details. If omitted (or otherwise falsey) then only default values will be used. This object can contain:
    - `commandTimeout` (**Number**) is the maximum number of milliseconds that we'll wait for a command on this sentinel to return. Default is 1500.
    - `createClient` (**Function**) is the function that will create the `RedisClient`s that are returned to you. By default, this library will try to require() the [node-redis](https://github.com/mranney/node_redis) library from the scope of your project automagically, and use our own version if yours is unavailable. If you want to do something more advanced, then you can use this argument to override the default. The function will be called with the arguments: `(port, host, options)`.
    - `customLogger` (**Function**) is the function used when logging messages. This function will only be called if `debugLogging` is set to true. Is called with a single string argument: `(message)`. Default: `console.log`.
    - `debugLogging` (**Boolean**) should be true if you want to see the noisy debug log. Default is `false`.
    - `outageRetryTimeout` (**Number**) is the number of milliseconds before trying again if ALL sentinels are down. If this number is negative, then we will simply emit an error instead of retrying. Default is 5000.
    - `randomizeSentinels` (**Boolean**) indicates whether the sentinels list should be shuffled before attempting connections. Default is true.
    - `redisOptions` (**Object**) is the object that is passed to createClient as options. Default is `{}`.
    - `refreshTimeout` (**Number**) is the number of milliseconds between attempts to fetch configurations from redis. Normally, we use events emitted to us through sentinel's pub/sub mechanism, but we always have this as a fallback should a message not make it to us. Default is 60000.
    - `timeout` (**Number**) is the connect timeout in milliseconds for connecting to a sentinel server. Default is 500.
    - `watchedNames` (**Array**) is an Array of the String names of Replicas to watch. Default is to watch everything that a sentinel is currently watching.

##### Event: "change"

Emitted whenever the connection information for a Replica Set has changed. Use this event to build RedisClient's for all of the servers that you're keeping track of. This will also be emitted during startup, so you can connect to Redis quickly.

Has arguments: `(name, replica)`
- `name` (**String**) is name of a watched Replica.
- `replica` (**RedisReplica**) is a `RedisReplica` object. Use this object to connect to masters or slaves. For the full API, [see below](#redisreplica).

##### Event: "error"

Emitted when there was an internal error, or if all redis-sentinels are down and retries have been disabled.

Has arguments: `(error)`
- `error` is an Error object. Big surprise. :)

##### Event: "event"

Emitted when an important event has happened. These are simply events that were passed to us through pub/sub in sentinel. Don't rely on these: some events may be omitted or duplicated in certain cases (such as when switching from one sentinel to another.) This is mostly useful for logging.

Has arguments: `(event_type, event_message)`
- `event_type` (**String**) is the event type, like `"+odown"`.
- `event_message` (**String**) describes what happened, like `"master main 127.0.0.1 9001 #quorum 4/3"`.


#### RedisReplica

This object stores the connection information to all object in a Replica Set, and gives you a bunch of useful methods for connecting to them with [node-redis](https://github.com/mranney/node_redis). It is provided in the `change` events.

##### Function: connectMaster()

Will create a connection to the master server in the Replica Set. Returns a brand-new `RedisClient`. Will be `null` if the server is currently objectively down, and there is no way to connect to it. (Uh-oh.)

##### Function: connectSlave()

Will create a connection to a random available slave in the Replica Set. A slave is unavailable if the Sentinel describes this server as being subjectively down. Returns a brand-new `RedisClient`. Will be `null` if no slaves are currently available.

##### Function connectAllSlaves()

Will create connections to all available slaves in a Replica Set. A slave is unavailable if the Sentinel describes this server as being subjectively down. Returns an Array of brand-new `RedisClient`s. Useful if you want to implement any round-robin load balancing, or what-have-you. This Array will be empty if no slaves are currently connectable.

##### Function: getMasterConfig()

Will fetch the connection information for the current master in the Replica Set. If the master is available, then returns an Object with the keys:
- `host` (**String**) is the hostname of the master. Should usually be an IP address.
- `port` (**Number**) is the port to connect to.

Returns `null` when master is unavailable.

##### Function: getSlaveConfig()

Will fetch the connection information for a random slave in the Replica Set. If a slave is available, then returns an Object with the keys:
- `host` (**String**) is the hostname of the slave. Should usually be an IP address.
- `port` (**Number**) is the port to connect to.

Returns `null` when all slaves are down.

##### Function: getAllSlaveConfigs()

Will fetch the connection information for all available slaves in the Replica Set. Returns an array of Objects, where each object has keys:
- `host` (**String**) is the hostname of the slave. Should usually be an IP address.
- `port` (**Number**) is the port to connect to.

### License

MIT


CHANGELOG.md
## Changelog

### 1.0.0 (*2016 Nov 20*)
 - Forked to `redsentinel` due to the package not being maintained.
 - Updated dependencies to latest.
 - Fixed a few bugs due to the version bump.
 - Dropped Node 0.8 support.

### 0.2.0 (*2015 Jul 2*)
 - Added API to get connection info without connecting.
 - Fixed a minor bug with error handling.

### 0.1.2 (*2014 Dec 14*)
 - Fixed write-to-global bug in the logging logic.
 - Added ability to use custom logging functions.

### 0.1.1 (*2014 Nov 26*)
 - NPM release.
 - Doc updates.
 - Sentinels array is now deduped.
 - Test fixes.

### 0.1.0 (*2014 Nov 18*)
 - Initial release. (Not on NPM though...)
 - Random sentinel connection to a sentinel group.
 - Master / Slave detection for replica sets.
 - Push notifications from sentinel on configuration changes.
 - Up / Down detection for RedisClient connections.
 - Can connect to masters, a random slave, or all slaves simultaneously.


src/redisWatcher.js
var util   = require('./util')
  , redis  = require('redis')
  , events = require('events');


// === Events Emitted =====
// error(err), refresh(), event()
// ========================

var RELOAD_EVENTS = [
  // These guys are useful, as slaves / masters toggle on and off:
  "+sdown", "-sdown", "+odown", "-odown",

  // These are useful to see when failovers happen according to the docs:
  "+reset-master", "switch-master",

  // These are useful to see when failovers happen via experimentation:
  "+role-change", "-role-change", "+switch-master"
];

var INTERESTING_EVENTS = [
  // Things the the outside public might be interested in:
  "+slave", "failover-end", "no-good-slave"
];

var USEFUL_EVENTS = RELOAD_EVENTS.concat(INTERESTING_EVENTS);


/**
 * Will create a RedisWatcher object, which opens the pub/sub channel
 * with a sentinel, and emits events when things should happen.
 *
 * @param {String} host    The host to connect to.
 * @param {Number} port    The port num.
 * @param {Number} config  The config object passed to the RedisSentinel.
 */
function RedisWatcher(host, port, config) {
  this.finalized = false;

  this.host = host;
  this.port = port;
  this.config = config;

  this.timeout = undefined;

  this._log = util.buildLogger("RedisWatcher");
  this._log.configure(config);

  var redis_config = {
    connect_timeout: config.timeout,
    retry_strategy: function () {}
  };

  // Hook ourselves into a redis client:
  this.client = config._testClient || redis.createClient(port, host, redis_config);

  this.client
    .on  ('error', this._handleClientError.bind(this))
    .on  ('end',   this._handleClientHangup.bind(this))
    .once('ready', this._handleConnectReady.bind(this));

  // Emit a starting refresh, but do it on next tick to give outside time to listen:
  this._log("Refreshing due to init");
  process.nextTick(this.emit.bind(this, 'refresh'));

  // Start the timer:
  this._resetTimer();
}

util.inherits(RedisWatcher, events.EventEmitter);


/**
 * Will halt the updater, closing all connections and freeing all assets.
 * @param  {Error} err  An error to emit when done. Optional.
 */
RedisWatcher.prototype.kill = function (err) {
  if (this.finalized) { return; }
  this.finalized = true;
  this.client.end(false);
  this.client.removeAllListeners();
  clearTimeout(this.timeout);
  this.timeout = undefined;
  this.emit("error", err);
  this._log("Closed connection to %s:%d", this.host, this.port);
};


RedisWatcher.prototype._handleClientError = function _handleClientError(err) {
  if (this.finalized) { return; }
  this.kill(err);
};


RedisWatcher.prototype._handleClientHangup = function _handleClientHangup() {
  this._log("Redis connection closed");
  this.kill();
};


RedisWatcher.prototype._handleConnectReady = function _handleConnectReady() {
  if (this.finalized) { return; }

  var watcher = this;

  // Start listening for interesting events:
  this.client.subscribe.apply(this.client, USEFUL_EVENTS);
  this.client.on('message', function (channel, message) {
    if (watcher.finalized) { return; }

    watcher._log("Got event:", channel, message);

    // Emit for clients:
    watcher.emit("event", channel, message);

    // Are we interested in this event?
    if (RELOAD_EVENTS.indexOf(channel) >= 0) {
      watcher._log("Refreshing due to", channel, "event");
      watcher.emit('refresh');
      watcher._resetTimer();
    }
  });
};


RedisWatcher.prototype._resetTimer = function _resetTimer() {
  if (this.finalized) { return; }

  var watcher = this;

  clearTimeout(this.timeout);

  this.timeout = setTimeout(function () {
    watcher._log("Refreshing due to timeout");
    watcher.emit('refresh');
    watcher._resetTimer();
  }, this.config.refreshTimeout);
};


module.exports = RedisWatcher;


src/redisConfigFetcher.js
var util   = require('./util')
  , redis  = require('redis')
  , events = require('events');


// === Events Emitted =====
// error(err), connected(), config(name, master, [slave])
// ========================


/**
 * A RedisConfigFetcher is responsible for connecting to a sentinel, and pulling
 * down a configuration for either specific replica names or all tracked replicas.
 *
 * @param {String} host    The host to connect to.
 * @param {Number} port    The port to connect to.
 * @param {Object} config  The config object passed to the RedisSentinel object.
 */
function RedisConfigFetcher(host, port, config) {
  this.finalized = false;

  var fetcher = this;

  this.host = host;
  this.port = port;
  this.config = config;
  this.cmd_to = config.commandTimeout;

  // State variables:
  this.is_fetching = false;
  this.needs_another = false;

  this._log = util.buildLogger("RedisConfigFetcher");
  this._log.configure(config);

  var redis_config = {
    connect_timeout: config.timeout, // Don't connect to down things
    no_ready_check:  1,              // Don't do a ready check, since we already do that
    retry_strategy: function () {}
  };

  var client = this.client = config._testClient || redis.createClient(port, host, redis_config);

  client.on('error', function (err) {
    fetcher.kill(err);
  });

  client.on('end', function () {
    fetcher._log("Redis connection closed");
    fetcher.kill();
  });

  client.once('ready', function () {
    if (fetcher.finalized) { return; }

    fetcher._log("Successfully connected to %s:%d", fetcher.host, fetcher.port);

    // Validate the sucker with an INFO check:
    util.timedCommand(client, fetcher.cmd_to, "INFO", function (err, result) {
      if (err || fetcher.finalized) {
        return fetcher.kill(err);
      }

      if (!RedisConfigFetcher._isInfoResponseValid(result)) {
        fetcher._log("Invalid response from sentinel");
        return fetcher.kill(new Error("Invalid response"));
      }

      fetcher.emit('connected');
    });
  });
}

util.inherits(RedisConfigFetcher, events.EventEmitter);


/**
 * Extremely simple and basic validation of the INFO response for a sentinel:
 * @param  {String}  info_data The string response from an INFO command.
 * @return {Boolean}           True IFF we're accepting it.
 */
RedisConfigFetcher._isInfoResponseValid = function _isInfoResponseValid(info_data) {
  if (!info_data || typeof info_data !== 'string') {
    return false;
  }

  if (info_data.indexOf("# Sentinel") < 0) {
    return false;
  }

  return true;
};


/**
 * Will take in the response from a "SENTINEL MASTERS" command, and return a parsed struct.
 *
 * @param  {String} role   The role of the server, like "Master". Only used for logging.
 * @param  {Array}  result The response from SENTINEL MASTERS.
 * @return {Object}       An object like: {"name": {ip: "...", ...}, ...} if ok. null otherwise
 */
RedisConfigFetcher.prototype._parseServerList = function _parseServerList(role, result) {
  var i, len, out = [];

  var ns = role + " list rejected:";

  if (!result || !Array.isArray(result)) {
    this._log(ns, "Bad input");
    return null;
  }

  for (i=0,len=result.length; i<len; i++) {
    var row = result[i], row_len, j, parsed_row = {};

    if (!Array.isArray(row) || row.length % 2 === 1) {
      this._log(ns, "Malformed row");
      return null;
    }

    for (j=0,row_len=row.length; j<row_len; j+=2) {
      var key = row[j], value = row[j+1];

      if (parsed_row.hasOwnProperty(key)) {
        this._log(ns, "Row had duplicate property");
        return null;
      }

      // Special-case for port entries:
      if (key === "port") {
        value = parseInt(value, 10);
        if (isNaN(value)) {
          this._log(ns, "Row has a non-numeric port");
          return null;
        }
      }

      // Special-case for flags:
      if (key === "flags") {
        value = value.split(",");
      }

      parsed_row[key] = value;
    }

    if (!parsed_row.hasOwnProperty("name")) {
      this._log(ns, "Row lacked a name property");
      return null;
    }

    out.push(parsed_row);
  }

  return out;
};


/**
 * Will build a lookup table for a list based on the "name" property of the contents.
 *
 * @param  {Array} array An array of objects, such as those returned from _parseServerList()
 * @return {Object}      A lookup table of objects in array from the name property to the elements.
 *                       Returns null on failure or duplicates.
 */
RedisConfigFetcher.prototype._buildLookup = function _buildLookup(array) {
  if (!array) {
    // Error probably happened in earlier parsing. Just fail w/o logging:
    return null;
  }

  var ns = "Failed to build lookup:";

  if (!Array.isArray(array)) {
    this._log(ns, "Non-array provided");
    return null;
  }

  var i, len = array.length, out = {};
  for (i=0; i<len; i++) {
    var item = array[i];

    if (!item.hasOwnProperty("name")) {
      this._log(ns, "Item #", i, "has no name");
      return null;
    }

    if (out.hasOwnProperty(item.name)) {
      this._log(ns, "Duplicate name in array:", item.name);
      return null;
    }

    out[item.name] = item;
  }

  return out;
};


/**
 * Will stop the Config Fetcher and release its assets. If an error
 * is provided, it will be emitted on the error channel.
 *
 * @param  {Error} err An optional error to emit.
 */
RedisConfigFetcher.prototype.kill = function kill(err) {
  if (this.finalized) { return; }
  this.finalized = true;

  this.client.end(false);
  this.client.removeAllListeners();

  this.emit('error', err || new Error());

  this._log("Connection to %s:%d closed.", this.host, this.port);
};


/**
 * Will get the Fetcher to fetch some new configs. If it already is,
 * it will fetch configs immediately after it's done. The new configs
 * will be emitted on the 'config' channel.
 */
RedisConfigFetcher.prototype.updateConfigs = function updateConfigs() {
  if (this.finalized) { return; }

  var fetcher = this;

  // If we're already doing a refresh, then flag that we need to do another and return:
  if (this.is_fetching) {
    this.needs_another = true;
    return;
  }

  this._log("Refreshing replica configurations...");

  // Set the state vars:
  this.needs_another = false;
  this.is_fetching = true;

  // Fetch the masters, so that we can select the ones with server configs:
  util.timedCommand(this.client, fetcher.cmd_to, "SENTINEL", ["masters"], function (err, res) {
    if (err || fetcher.finalized) { return fetcher.kill(err); }

    // Fetch all masters as a lookup from name => parsed config object.
    var all_masters = fetcher._buildLookup(fetcher._parseServerList("Master", res));
    if (!all_masters) { return fetcher.kill(new Error("Parsing master list failed")); }

    // Get the list of watched replicas, and select only those ones from the master list:
    var tracked_names   = fetcher.config.watchedNames || Object.keys(all_masters)
      , tracked_masters = {};

    tracked_names.forEach(function (name) {
      if ( !all_masters.hasOwnProperty(name) ) {
        fetcher._log("Warning: Replica name in watchedNames but not on Sentinel:", name);
        return;
      }
      tracked_masters[name] = all_masters[name];
    });

    // Items may have been removed from tracked_names to tracked_masters, if the name isn't
    // tracked by the sentinel:
    tracked_names = Object.keys(tracked_masters);

    // Now pull in all slave configurations:
    util.async.each(
      tracked_names,
      function fetchSlaves(name, done) {
        if (fetcher.finalized) { return; }

        util.timedCommand(fetcher.client, fetcher.cmd_to, "SENTINEL", ["slaves", name], function (err, slaves) {
          if (err || fetcher.finalized) {
            return fetcher.kill(err);
          }

          // Parse the slave data:
          var parsed_slaves = fetcher._parseServerList("Slave", slaves);
          if (!parsed_slaves) { return fetcher.kill(new Error("Parsing slave list failed")); }

          // Pass this along to the outside:
          fetcher.emit('config', name, tracked_masters[name], parsed_slaves);
          done();
        });
      },
      function allDone(err) {
        if (err || fetcher.finalized) { return fetcher.kill(err); }

        // Ok, done running:
        fetcher.is_fetching = false;

        fetcher._log("Done fetching configs");

        // Do we already have another refresh scheduled?
        if (fetcher.needs_another) {
          return process.nextTick(fetcher.updateConfigs.bind(fetcher));
        }
      }
    );
  });
};


module.exports = RedisConfigFetcher;


src/redisSentinel.js
var events = require('events')
  , redis = require('redis')
  , RedisReplica = require('./redisReplica')
  , RedisConfigFetcher = require('./redisConfigFetcher')
  , RedisWatcher = require('./redisWatcher')
  , util = require('./util');


// This is passed into the RedisReplica, so that the client redis objects are their own:
var client_redis = null;

// Uncomment this to debug the redis protocol stuffs:
// redis.debug_mode = true;


function RedisSentinel(sentinels, options) {
  if (!(this instanceof RedisSentinel)) {
    return new RedisSentinel(sentinels, options);
  }

  // Add in the things that will be asynchronously populated later:
  this.fetcher = null;
  this.watcher = null;
  this.replicas = {};

  // Keep track of death:
  this.terminated = false;

  // Simple validation:
  RedisSentinel._validateSentinelList(sentinels);

  // Extend the default values with the custom parameters:
  var defaults = {
    commandTimeout:     1500,
    createClient:       client_redis && client_redis.createClient || redis.createClient,
    customLogger:       null,
    debugLogging:       false,
    outageRetryTimeout: 5000,
    randomizeSentinels: true,
    redisOptions:       {},
    refreshTimeout:     60000,
    timeout:            500,
    watchedNames:       null,
  };

  this.options = util._.extend(defaults, options);

  // Make sure that the outside can't inject a test client:
  delete this.options._testClient;

  this._log = util.buildLogger("RedisSentinel");
  this._log.configure(this.options);

  this._log("Initialization started. Has client redis:", !!client_redis);

  // Store connection info for all the sentinels:
  this.sentinels = sentinels.map(function (conf) {
    var host = String(conf.host)
      , port = parseInt(conf.port, 10) || 26379;

    return { host: host, port: port };
  });

  this._dedupeSentinels();

  // Start this sucker on its way:
  this._connectSentinel();
}

// Inherit from EventEmitter, so we can emit stuff:
util.inherits(RedisSentinel, events.EventEmitter);


/**
 * Will do basic validation on a sentinel list. Throws if there was a problem.
 *
 * @param  {Array} sentinels The sentinel list, from the user.
 */
RedisSentinel._validateSentinelList = function _validateSentinelList(sentinels) {

  // First, check type:
  if (!sentinels || !Array.isArray(sentinels)) {
    throw new TypeError("First argument needs to be an array of sentinel config objects");
  }

  // Check array length:
  if (!sentinels.length) {
    throw new TypeError("Sentinels array shouldn't be empty");
  }

  // Now go through each item in the list, and make sure that there's the necessary info:
  sentinels.forEach(function (item, idx) {
    if (!item || typeof item !== 'object') {
      throw new TypeError("Item #" + idx + " in sentinels array isn't an object");
    }
    if (!item.host || typeof item.host !== 'string') {
      throw new Error("Item #" + idx + " in sentinels array doesn't have a correct host property");
    }
  });
};


/**
 * Will try to connect to each item in the sentinels array, in order, until it is
 * successful.
 */
RedisSentinel.prototype._connectSentinel = function _connectSentinel() {
  var sentinel = this;

  this._log("Starting sentinel connection...");

  // Make sure we are iterating over the sentinels in a random order:
  if (this.options.randomizeSentinels) {
    util.shuffleArray(this.sentinels);
  }

  // Call this when shit fucks up:
  var _abortAndReconnect = util._.once(sentinel._handleErrorAndReconnect.bind(sentinel));

  // Try them all!
  util.async.forEachSeries(
    this.sentinels,
    function withEachSentinel(conf, next) {

      var fetcher = sentinel.fetcher = new RedisConfigFetcher(conf.host, conf.port, sentinel.options)
        .once('error', _onError)
        .once('connected', _onConnect);

      // Errors at this stage mean that we will move to the next sentinel:
      function _onError() {
        fetcher.removeListener('connected', _onConnect);
        if (sentinel.terminated) { return; }
        next();
      }

      // Else, cool. Re-hook this thing up in a more stable way, and ask it for a config:
      function _onConnect() {
        fetcher.removeListener('error', _onError);
        if (sentinel.terminated) { return; }

        // Hook up the event handlers for the long run:
        fetcher
          .on('error', _abortAndReconnect)
          .on('config', _onGetReplInfo);

        // Now, create a watcher to poke this sucker along:
        sentinel.watcher = new RedisWatcher(conf.host, conf.port, sentinel.options)
          .on('error', _abortAndReconnect)
          .on('event', _passAlongEvent)
          .on('refresh', fetcher.updateConfigs.bind(fetcher));

        function _passAlongEvent(channel, msg) {
          if (sentinel.terminated) { return; }
          sentinel.emit('event', channel, msg);
        }

        function _onGetReplInfo(name, master, slaves) {
          if (sentinel.terminated) { return; }

          // Create a new RedisReplica if we've never seen this name before:
          if (!sentinel.replicas.hasOwnProperty(name)) {
            sentinel.replicas[name] = new RedisReplica(name, sentinel.options.createClient, sentinel.options.redisOptions);
          }

          // Load the master / slave configs into the struct:
          var repl           = sentinel.replicas[name]
            , master_changed = repl._loadMasterConfig(master)
            , slaves_changed = repl._loadSlaveConfigs(slaves)
            , has_changed    = master_changed || slaves_changed;

          sentinel._log("Repl configuration set:", repl.toString());

          // Emit the repl if things are different:
          if (has_changed) {
            sentinel._log("Repl", name, "has changed");
            sentinel.emit('change', name, repl);
          }
        }
      }
    },
    function (err) {
      if (sentinel.terminated) { return; }
      if (err) {
        sentinel.emit("error", err);
        return;
      }

      // We made it through all endpoints. What do??
      if (sentinel.options.outageRetryTimeout < 0) {
        // Stop, and emit error:
        sentinel.emit("error", new Error("Could not connect to a sentinel. *<:'O("));
        return;

      } else {
        // Loop around and try again in a few seconds:
        sentinel._log("All sentinels down. Pausing before retry...");
        setTimeout(sentinel._connectSentinel.bind(sentinel), sentinel.options.outageRetryTimeout);
      }
    }
  );
};


RedisSentinel.prototype._handleErrorAndReconnect = function _handleErrorAndReconnect(err) {
  if (this.terminated) { return; }

  // Log the error:
  this._log("Error encountered:", err);

  // Kill our redis clients:
  if (this.fetcher) { this.fetcher.kill(); }
  if (this.watcher) { this.watcher.kill(); }
  this.fetcher = this.watcher = null;

  // Trigger a re-connect:
  this._connectSentinel();
};


RedisSentinel.prototype._dedupeSentinels = function _dedupeSentinels() {
  // Dedupe the list based on "host(lowercase):port":
  this.sentinels = util._.uniqBy(this.sentinels, function (conf) {
    return conf.host.toLowerCase() + ":" + conf.port;
  });
};


RedisSentinel.prototype.kill = function kill() {
  if (this.terminated) { return; }
  this.terminated = true;

  this._log("Termination triggered from outside");

  // Kill our redis clients:
  if (this.fetcher) { this.fetcher.kill(); }
  if (this.watcher) { this.watcher.kill(); }
  this.fetcher = this.watcher = null;
};


// Exports:
module.exports.RedisSentinel = RedisSentinel;

// Internal API for passing along the client's Redis library:
module.exports.setClientRedis = function setClientRedis(mod) {
  client_redis = mod;
};
